File: cpp/generic/examples/loadModel.cpp
// Must be included before ASL, which
// is included by cplex_interface and gurobi_interface
#include "gurobi_c.h"
#include "ilcplex/cplex.h"

#include "cplex_interface.h"

#include "gurobi_interface.h"

#include "xpress_interface.h"

#include "simpleapi/simpleApi.h"
#include "test-config.h" // for MODELS_DIR

const char* MODELNAME = "tsp.nl";
double doStuff(ampls::AMPLModel& m, const char *name) 
{
  // Start the optimization process
  m.optimize();
  printf("\n");
  // Get the generic status
  ampls::Status s = m.getStatus();
  switch (s)
  {
  case ampls::Status::Optimal:
    printf("Optimal ");
    break;
  case ampls::Status::Infeasible:
    printf("Infeasible ");
    break;
  case ampls::Status::Unbounded:
    printf("Unbounded ");
    break;
  default:
    printf("Status (%d) ", s);
  }
  // Get the objective value
  double obj = m.getObj();
  printf("solution with %s=%f\n", name, obj);
  

  // Get the solution vector and count the non-zeros
  std::vector<double> solution = m.getSolutionVector();
  int nnz = 0;
  for (int i = 0; i < solution.size(); i++)
    if (solution[i] != 0) nnz++;
  printf("Number of non zeroes = %d\n", nnz);
  // Write the AMPL sol file
  char BUFFER[1024];
  sprintf(BUFFER, "%s-%s.sol", m.getFileName().c_str(), name);
  printf("Writing solution file to: %s\n", BUFFER);
  m.writeSol(BUFFER);
  return obj;
}

int main(int argc, char** argv) {
  char buffer[80];
  strcpy(buffer, MODELS_DIR);
  strcat(buffer, MODELNAME);

  // Load a model using gurobi driver
  ampls::GurobiDrv gurobi;
  ampls::GurobiModel g = gurobi.loadModel(buffer);
  // Use it as generic model
  doStuff(g, "gurobi");

  // Load a model using CPLEX driver
  ampls::CPLEXDrv cplex;
  ampls::CPLEXModel c = cplex.loadModel(buffer);
  // Use it as generic model
  doStuff(c, "cplex");

  // Load a model using CPLEX driver
  ampls::XPRESSDrv xpress;
  ampls::XPRESSModel x = xpress.loadModel(buffer);
  // Use it as generic model
  doStuff(x, "xpress");
  return 1;
 
}

File: python/simpleapi/examples/loadModel.py
import sys
import amplpy_simpleapi_swig as ampls

#MODEL = "d:\\model.nl"
MODEL = "D:/Development/ampl/solvers-private/build/vs64/bin/nlfileexport.nl"

statuses= {
    ampls.Status.Optimal :"optimal", 
    ampls.Status.Infeasible :"infeasible", 
    ampls.Status.Interrupted :"interrupted", 
    ampls.Status.LimitIteration :"limititeration", 
    ampls.Status.LimitNode :"limitnode", 
    ampls.Status.LimitSolution:"limitsolution",
    ampls.Status.LimitTime:"limittime",
    ampls.Status.Unbounded:"unbounded",
    ampls.Status.Unknown:"unknown",
    ampls.Status.NotMapped:"not mapped"
  }

# Generic function doing the optimization and reading the results
def doStuff(a: ampls.AMPLModel):
  # Optimize with default settings
  a.optimize()
  # Print the status (using the map above to have a description)
  print("Model status ", statuses[a.getStatus()])
  # Print the objective function
  print("Objective: ",a.getObj())
  # Get the solution vector
  d = a.getSolutionVector()
  # Print the non-zeroes, by getting a map
  # Note that this will only work if the col file has been generated
  map = a.getVarMapInverse()
  nonzeroes = [(map[index], d[index]) for index in map if d[index] != 0]
  for (name, value) in nonzeroes:
    print("{} = {}".format(name, value))

# Actual script
cplex = ampls.CPLEXDrv() # Create the driver
cplexModel = cplex.loadModel(MODEL) # Load the model
doStuff(cplexModel) # Call the (generic) function

gurobi = ampls.GurobiDrv() # Create the driver
gurobiModel = gurobi.loadModel(MODEL) # Load the model
doStuff(gurobiModel) # Call the (generic) function
